crates/libsolv_rs/src/solver/mod.rs

```diff
@@ -0,0 +1,1399 @@
+                break;
+            };
+
+            {
+                let solvable = self
+                    .pool
+                    .resolve_solvable_inner(conflicting_solvable)
+                    .display();
+                println!(
+                    "=== Propagation conflicted: could not set {solvable} to {attempted_value}"
+                );
+                println!(
+                    "During unit propagation for clause: {:?}",
+                    self.clauses[conflicting_clause.index()].debug(&self.pool)
+                );
+
+                let decision = self
+                    .decision_tracker
+                    .stack()
+                    .iter()
+                    .find(|d| d.solvable_id == conflicting_solvable)
+                    .unwrap();
+                println!(
+                    "Previously decided value: {}. Derived from: {:?}",
+                    !attempted_value,
+                    self.clauses[decision.derived_from.index()].debug(&self.pool),
+                );
+            }
+
+            if level == 1 {
+                println!("=== UNSOLVABLE");
+                for decision in self.decision_tracker.stack() {
+                    let clause = &self.clauses[decision.derived_from.index()];
+                    let level = self.decision_tracker.level(decision.solvable_id);
+                    let action = if decision.value { "install" } else { "forbid" };
+
+                    if let Clause::ForbidMultipleInstances(..) = clause.kind {
+                        // Skip forbids clauses, to reduce noise
+                        continue;
+                    }
+
+                    println!(
+                        "* ({level}) {action} {}. Reason: {:?}",
+                        self.pool
+                            .resolve_solvable_inner(decision.solvable_id)
+                            .display(),
+                        clause.debug(&self.pool),
+                    );
+                }
+
+                return Err(self.analyze_unsolvable(conflicting_clause));
+            }
+
+            let (new_level, learned_clause_id, literal) =
+                self.analyze(level, conflicting_solvable, conflicting_clause);
+            level = new_level;
+
+            println!("=== Backtracked to level {level}");
+
+            // Optimization: propagate right now, since we know that the clause is a unit clause
+            let decision = literal.satisfying_value();
+            self.decision_tracker
+                .try_add_decision(
+                    Decision::new(literal.solvable_id, decision, learned_clause_id),
+                    level,
+                )
+                .expect("bug: solvable was already decided!");
+            println!(
+                "=== Propagate after learn: {} = {decision}",
+                self.pool
+                    .resolve_solvable_inner(literal.solvable_id)
+                    .display()
+            );
```

Comment:
Just as a future idea - this would probably the place to check for a time-budget, right?

Reply:
You mean giving up after a given amount of time? In my eyes the best place for that would be `Solver::resolve_dependencies`, which is using `Solver::set_propagate_learn`

Reply:
Yeah, like setting a limit of 1 sec, or 10'000 iterations ...